### PR TITLE
Add support for customizing transaction start - fixes #432

### LINKED ISSUE #432: Support for setting Transaction Isolation
Reference: 

https://vertx.io/docs/apidocs/io/vertx/ext/sql/SQLOptions.html#getTransactionIsolation--
https://vertx.io/docs/apidocs/io/vertx/ext/sql/TransactionIsolation.html

### PR DESCRIPTION
**Pull request by BillyYccc**
Motivation:

fixes #432 

Conformance:

Your commits should be signed and you should have signed the Eclipse Contributor Agreement as explained in https://github.com/eclipse/vert.x/blob/master/CONTRIBUTING.md
Please also make sure you adhere to the code style guidelines: https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines

### CODE REVIEW

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/TransactionImpl.java [hunk 1]
```diff
@@ -62,9 +63,9 @@
     }
   }
 
-  Future<Transaction> begin() {
+  Future<Transaction> begin(String startTxSql) {
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/TransactionImpl.java:66]: startTxSql should be replace by an enum as each database might use something different to do this

Comment (vietj) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/TransactionImpl.java:66]: also user don't want to know about the SQL for this :-)

**Comment by BillyYccc** [comment]
I have rethinked and add a prototype here(the code may be a little messy but it can show the basic design)

The primary idea is to provide a simple API to configure transaction options at the start the transaction, and it would only support those common defined behaviors.
So now we have an unified API to configure transaction isolation levels and transaction access modes, specific database sql modifiers are not included in the API such as MySQL `WITH CONSISTENT SNAPSHOT` and Postgres `DEFERRABLE`.

Even though we have the same API, the sql to start a transaction in different databases is different, therefore we have to build the begin tx SQL at the codec level for different databases.
Since MySQL it does not support specifying the isolation level in the start tx sql like Postgres, a SQL `SET TRANSACTION ISOLATION LEVEL <...>` needs to be queried before the transaction start so that the isolation level is correctly set. I think we would go in the same way to solve this kind of issue for DB2/SQL Server as well.

If you're fine with this API, I will proceed to clean the code and work at this.

**Comment by gavinking** [comment]
LGTM.

The ANSI standard options are `ISOLATION LEVEL READ UNCOMMITTED`, `ISOLATION LEVEL READ COMMITTED`, `ISOLATION LEVEL REPEATABLE READ`, `ISOLATION LEVEL SERIALIZABLE`, and `READ ONLY`, `READ WRITE`. 

ANSI SQL also defines `DIAGNOSTICS SIZE`, but I've never seen that supported in any database.

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/TransactionOptions.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package io.vertx.sqlclient;
+
+import io.vertx.codegen.annotations.DataObject;
+
+//TODO codegen
+public class TransactionOptions {
+
+  public static final TransactionOptions DEFAULT_TX_OPTIONS = new TransactionOptions();
+
+  private TransactionIsolationLevel transactionIsolationLevel;
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/TransactionOptions.java:10]: simply _isolationLevel_

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/TransactionOptions.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  private TransactionAccessMode transactionAccessMode;
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/TransactionOptions.java:11]: simply _accessMode_

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/command/StartTxCommand.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package io.vertx.sqlclient.impl.command;
+
+import io.vertx.sqlclient.TransactionOptions;
+
+public class StartTxCommand<R> extends TxCommand<R> {
+  public TransactionOptions transactionOptions;
+
+  public StartTxCommand(Kind kind, R result, TransactionOptions transactionOptions) {
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/command/StartTxCommand.java:8]: rather here directly the access mode and the isolation level as _ TransactionOptions_ is mutable

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/Pool.java [hunk 5]
```diff
@@ -152,6 +153,14 @@ static Pool pool(Vertx vertx, SqlConnectOptions connectOptions, PoolOptions pool
    */
   <T> Future<T> withTransaction(Function<SqlClient, Future<T>> function);
 
+  @GenIgnore
+  // FIXME codegen
```
Comment (vietj) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/Pool.java:157]: what is the codegen error here ?

**Comment by vietj** [comment]
That makes some new classes in the top level package, I think we should have them in a `transaction` child package instead

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/Pool.java (hunk 5 above)
Comment (BillyYccc) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/Pool.java:157]: it should be ok, I just don't add codegen annotations for a POC prototype, will add it later

**Pull request closed without merging**